File: StandardLibrary/runtimes/net/Extern/OsLang.cs
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

using icharseq = Dafny.ISequence<char>;
using charseq = Dafny.Sequence<char>;

namespace OsLang
{
    public partial class __default
    {
        // returns one of : MacOS, Windows, Unix, Other
        public static icharseq GetOsShort()
        {
            // PlatformID.MacOSX doesn't actually work
            String osDesc = System.Runtime.InteropServices.RuntimeInformation.OSDescription;
            if (osDesc.Contains("Darwin"))
            {
                return charseq.FromString("MacOS");
            }
            OperatingSystem os = Environment.OSVersion;
            PlatformID pid = os.Platform;
            switch (pid)
            {
                case PlatformID.Win32NT:
                case PlatformID.Win32S:
                case PlatformID.Win32Windows:
                case PlatformID.WinCE:
                    return charseq.FromString("Windows");
                case PlatformID.Unix:
                    return charseq.FromString("Unix");
                case PlatformID.MacOSX:
                    // This doesn't actually happen on MacOS
                    return charseq.FromString("MacOS");
                default:
                    return charseq.FromString("Other");
            }
        }

        // returns one of : Java, Dotnet, Go, Python, Rust
        public static icharseq GetLanguageShort()
        {
            return charseq.FromString("Dotnet");
        }

        // returns a human readable, language specific, unstable over time, description of the OS
        public static icharseq GetOsLong()
        {
            return charseq.FromString(System.Runtime.InteropServices.RuntimeInformation.OSDescription);
        }

        // returns a human readable, language specific, unstable over time, description of the Language
        public static icharseq GetLanguageLong()
        {
            return charseq.FromString("Dotnet " + Environment.Version.ToString());
        }
    }
}

File: StandardLibrary/runtimes/python/src/smithy_dafny_standard_library/internaldafny/extern/OsLang.py
import os
import _dafny
import platform
import sys

import smithy_dafny_standard_library.internaldafny.generated.OsLang

# Extend generated class with our externs
class default__(smithy_dafny_standard_library.internaldafny.generated.OsLang.default__):
    def GetOsShort():
        sys = platform.system()
        if sys == "Darwin":
            return _dafny.Seq(_dafny.string_of("MacOS"))
        if sys == "Linux":
            return _dafny.Seq(_dafny.string_of("Unix"))
        return _dafny.Seq(_dafny.string_of(sys))

    def GetOsLong():
        return _dafny.Seq(_dafny.string_of(platform.platform()))

    def GetLanguageShort():
        return _dafny.Seq(_dafny.string_of("Python"))

    def GetLanguageLong():
        return _dafny.Seq(_dafny.string_of("Python " + sys.version))

# Export externs
smithy_dafny_standard_library.internaldafny.generated.OsLang.default__ = default__
